modules/querying/pages/func/query-user-defined-functions.adoc

```diff
@@ -1,46 +1,67 @@
 
-This page introduces the process to define a query UDF. Once defined, the new functions will be added into GSQL automatically next time GSQL is executed.
+This page introduces the process to define a query UDF. Once defined, the new functions are added into GSQL automatically the next time GSQL is executed.
+
+Query UDFs are user-editable in the form of an `.hpp` file locally on your machine. These get uploaded and interpreted by the server through a GSQL command. Creating, editing, or removing Query UDFs is done by downloading them from the server, editing the file, and reuploading.
 
 == Define a query UDF
 
 Below are the steps to add a Query UDF to GSQL:
 
 === Step 1: Download current query UDF file
 
-Use the `GET ExprFunctions` command in GSQL to download the current UDF file to any location on your machine. The file and the directores will be created if they do not exist, and the file must end with the file extention `.hpp`:
+Use the `GET ExprFunctions` command in GSQL to download the current list of functions into a local file. The path can be absolute or relative to your current directory, but the file extension must be `.hpp`:
 
-[source,text]
+[source,gsql]
 ----
 GSQL > GET ExprFunctions TO "/example/path/to/ExprFunctions.hpp"
+GSQL > GET ExprFunctions TO "./ExprFunctions.hpp"
+
 ----
 
 If your query UDF requires a user-defined struct or helper function, also use the `GET ExprUtil` command to download the current `ExprUtil` file:
 
-[source,text]
+[source,gsql]
 ----
 GSQL > GET ExprUtil TO "/example/path/ExprUtil.hpp"
 ----
 
 === Step 2: Define C{pp} function
 
-Define the C{pp} function inside the `UDIMPL` namespace inside of the UDF file you just downloaded in Step 1. The definition of the function should include the keyword `inline`. Only `bool`, `int`, `float`, `double`, and `string` (NOT `std::string`) are allowed as the return value type and the function argument type. However, any C{pp} type is allowed inside a function body.
+Define the C{pp} function inside the UDF file you just downloaded in Step 1. The definition of the function should include the keyword `inline`.
+
+Below is a list of data types and where each is allowed in a function.
+
+.Allowed Data Types
+|===
+|Data Type | Argument | Return | Function Body
+| `bool` | Yes | Yes | Yes
+| `int` | Yes | Yes | Yes
+| `float` | Yes | Yes | Yes
+| `double` | Yes | Yes | Yes
+| `string` | Yes | Yes | Yes
+| `std::string` | *No* | *No* | Yes
+| xref:accumulators.adoc[Accumulators] | Yes | Yes | Yes
+
+| All other C{pp} data types | *No* | *No* | Yes
+|===
 
 If the function requires a user-defined struct or helper function, define it in the `ExprUtil` file you downloaded in Step 1.
 
-Below is an example of a query UDF definition:
+Below is an example of a query UDF definition that includes two separate functions. The first function returns `true` if the passed value is greater than 3, and the second reverses a string.
```

Comment:
"Below is an example of a query UDF definition that includes two separate functions". I believe this is a misuse of _UDF_.  That file contains two UDFs, not one. They aren't even related. It just happens that the user wants to add them at the same time. UDF literally means User-Defined Function, so it makes no sense to say, "This user-defined function contains two functions" (unless they are subfunctions).